async functions compatibility

It would be great if we could use log-derive with async functions. Adding `logfn` to an async function gives me the following:

```
error[E0728]: `await` is only allowed inside `async` functions and blocks
  --> src/models/video/mod.rs:94:9
   |
44 | #[logfn(ok = "INFO", err = "ERROR")]
   | ------------------------------------ this is not `async`
...
94 |         task::sleep(std::time::Duration::from_millis(500)).await;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ only allowed inside `async` functions and blocks
```

Source Snippet:
```
#[logfn(ok = "INFO", err = "ERROR")]
pub async fn create_video_sdp() {
// ...
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -10,7 +10,10 @@ rust:
   - nightly-2019-11-25-x86_64-unknown-linux-gnu
 script:
   - cargo build --verbose
-  - if [ ${TRAVIS_RUST_VERSION} != "1.32.0" ]; then
+  - if [ ${TRAVIS_RUST_VERSION} != "nightly-2019-05-22-x86_64-unknown-linux-gnu" ] &&
+       [ ${TRAVIS_RUST_VERSION} != "nightly-2019-08-01-x86_64-unknown-linux-gnu" ] &&
+       [ ${TRAVIS_RUST_VERSION} != "1.32.0" ];
+    then
       cargo test --features async,async_test --verbose;
     else
       cargo test --verbose;

--- a/.travis.yml
+++ b/.travis.yml
@@ -14,7 +14,7 @@ script:
        [ ${TRAVIS_RUST_VERSION} != "nightly-2019-08-01-x86_64-unknown-linux-gnu" ] &&
        [ ${TRAVIS_RUST_VERSION} != "1.32.0" ];
     then
-      cargo test --features async,async_test --verbose;
+      cargo test --features async_test --verbose;
     else
       cargo test --verbose;
     fi

--- a/.travis.yml
+++ b/.travis.yml
@@ -10,7 +10,11 @@ rust:
   - nightly-2019-11-25-x86_64-unknown-linux-gnu
 script:
   - cargo build --verbose
-  - cargo test --verbose
+  - if [ ${TRAVIS_RUST_VERSION} != "1.32.0" ]; then
+      cargo test --features async,async_test --verbose;
+    else
+      cargo test --verbose;
+    fi
   - cargo test --examples
 
 before_script:

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,12 +14,17 @@ include = [
     "Cargo.toml",
 ]
 
+[features]
+async = []
+async_test = ["async", "futures-executor"]
+
 [dependencies]
 darling = "0.10.0"
 proc-macro2 = "1.0.3"
 #syn = { version = "0.15", features = ["full", "extra-traits"] } # -> For development
 syn = { version = "1.0.5", features = ["full"] }
 quote = "1.0.2"
+futures-executor = { version = "0.3.5", optional = true } # Can't have optional dev-dependency
 
 [dev-dependencies]
 simplelog = "0.7"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,8 +15,7 @@ include = [
 ]
 
 [features]
-async = []
-async_test = ["async", "futures-executor"]
+async_test = ["futures-executor"]
 
 [dependencies]
 darling = "0.10.0"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -99,8 +99,8 @@ use quote::{quote, ToTokens};
 
 use syn::punctuated::Punctuated;
 use syn::{
-    parse_macro_input, spanned::Spanned, token, AttributeArgs, Expr, ExprBlock, ExprClosure, FnArg, Ident, ItemFn, Meta, NestedMeta,
-    Pat, Result, ReturnType, Stmt, Type, TypePath,
+    parse_macro_input, spanned::Spanned, token, AttributeArgs, Expr, ExprAsync, ExprAwait, ExprBlock, ExprCall, ExprClosure,
+    ExprParen, FnArg, Ident, ItemFn, Meta, NestedMeta, Pat, Result, ReturnType, Stmt, Type, TypePath,
 };
 
 struct FormattedAttributes {
@@ -277,20 +277,43 @@ fn get_logger_token(att: &Ident) -> TokenStream {
     quote!(log::Level::#att_str)
 }
 
-fn make_closure(original: &ItemFn) -> ExprClosure {
-    let body =
-        Box::new(Expr::Block(ExprBlock { attrs: Default::default(), label: Default::default(), block: *original.block.clone() }));
-
-    ExprClosure {
-        attrs: Default::default(),
-        asyncness: Default::default(),
-        movability: Default::default(),
-        capture: Some(token::Move { span: original.span() }),
-        or1_token: Default::default(),
-        inputs: Default::default(),
-        or2_token: Default::default(),
-        output: ReturnType::Default,
-        body,
+fn make_closure(original: &ItemFn) -> Expr {
+    match original.sig.asyncness {
+        Some(asyncness) => Expr::Await(ExprAwait {
+            attrs: Default::default(),
+            await_token: Default::default(),
+            dot_token: Default::default(),
+            base: Box::new(syn::Expr::Async(ExprAsync {
+                attrs: Default::default(),
+                capture: Some(token::Move { span: original.span() }),
+                block: *original.block.clone(),
+                async_token: asyncness,
+            })),
+        }),
+        None => Expr::Call(ExprCall {
+            attrs: Default::default(),
+            args: Default::default(),
+            paren_token: Default::default(),
+            func: Box::new(syn::Expr::Paren(ExprParen {
+                attrs: Default::default(),
+                paren_token: Default::default(),
+                expr: Box::new(syn::Expr::Closure(ExprClosure {
+                    attrs: Default::default(),
+                    asyncness: Default::default(),
+                    movability: Default::default(),
+                    capture: Some(token::Move { span: original.span() }),
+                    or1_token: Default::default(),
+                    inputs: Default::default(),
+                    or2_token: Default::default(),
+                    output: ReturnType::Default,
+                    body: Box::new(Expr::Block(ExprBlock {
+                        attrs: Default::default(),
+                        label: Default::default(),
+                        block: *original.block.clone(),
+                    })),
+                })),
+            })),
+        }),
     }
 }
 
@@ -300,7 +323,7 @@ fn replace_function_headers(original: ItemFn, new: &mut ItemFn) {
     new.block = block;
 }
 
-fn generate_function(closure: &ExprClosure, expressions: FormattedAttributes, result: bool) -> Result<ItemFn> {
+fn generate_function(closure: &Expr, expressions: FormattedAttributes, result: bool) -> Result<ItemFn> {
     let FormattedAttributes { ok_expr, err_expr, log_ts, contained_ok_or_err } = expressions;
     let result = result || contained_ok_or_err;
     let code = if log_ts {
@@ -308,7 +331,7 @@ fn generate_function(closure: &ExprClosure, expressions: FormattedAttributes, re
             quote! {
                 fn temp() {
                     let instant = std::time::Instant::now();
-                    let result = (#closure)();
+                    let result = #closure;
                     let ts = instant.elapsed();
                     result.map(|result| { #ok_expr; result })
                         .map_err(|err| { #err_expr; err })
@@ -318,7 +341,7 @@ fn generate_function(closure: &ExprClosure, expressions: FormattedAttributes, re
             quote! {
                 fn temp() {
                     let instant = std::time::Instant::now();
-                    let result = (#closure)();
+                    let result = #closure;
                     let ts = instant.elapsed();
                     #ok_expr;
                     result
@@ -328,7 +351,7 @@ fn generate_function(closure: &ExprClosure, expressions: FormattedAttributes, re
     } else if result {
         quote! {
             fn temp() {
-                let result = (#closure)();
+                let result = #closure;
                 result.map(|result| { #ok_expr; result })
                     .map_err(|err| { #err_expr; err })
             }
@@ -336,7 +359,7 @@ fn generate_function(closure: &ExprClosure, expressions: FormattedAttributes, re
     } else {
         quote! {
             fn temp() {
-                let result = (#closure)();
+                let result = #closure;
                 #ok_expr;
                 result
             }

--- a/tests/test_async.rs
+++ b/tests/test_async.rs
@@ -1,0 +1,27 @@
+#![cfg(feature = "async_test")]
+
+use log_derive::logfn;
+
+#[logfn(INFO)]
+async fn async_function(ok: bool) -> Result<&'static str, &'static str> {
+    if ok {
+        return Ok("async Ok");
+    } else {
+        return Err("async Err");
+    }
+}
+
+#[test]
+fn async_works() {
+    init_logger();
+
+    futures_executor::block_on(async {
+        assert_eq!(async_function(true).await, Ok("async Ok"));
+        assert_eq!(async_function(false).await, Err("async Err"));
+    })
+}
+
+fn init_logger() {
+    // Run with `cargo test -- --test-threads=1 --nocapture`
+    let _ = simplelog::SimpleLogger::init(simplelog::LevelFilter::Info, simplelog::Config::default());
+}

